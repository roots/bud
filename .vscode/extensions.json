{
  "recommendations": [
    "arcanis.vscode-zipfs",
    "dbaeumer.vscode-eslint",
    "esbenp.prettier-vscode",
    "ms-vscode-remote.remote-containers",
    "ZixuanChen.vitest-explorer"
  ]
}
