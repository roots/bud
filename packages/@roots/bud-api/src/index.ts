import {Api} from './Api'
import * as methods from './methods'
export {Api, methods}
