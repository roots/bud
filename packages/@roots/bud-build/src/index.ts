// Copyright (c) Roots Foundation, LLC. All rights reserved.
// Licensed under the MIT license.

/**
 * Compiler configuration builder
 *
 * @see https://roots.io/bud
 * @see https://github.com/roots/bud
 *
 * @remarks
 * - 💁 Composable - Build exceptional applications with a modular, configurable build system
 *
 * - 💪 Modern - Modern framework written in TypeScript with an expressive API
 *
 * - 🌱 Easy - Low bundle size and fast build times
 *
 *  @packageDocumentation
 */

export {Build} from './Build'
export {Loader} from './Loader'
export {Item} from './Item'
export {Rule} from './Rule'

import * as items from './Build/items'
import * as loaders from './Build/loaders'
import * as rules from './Build/rules'
export {items, loaders, rules}
