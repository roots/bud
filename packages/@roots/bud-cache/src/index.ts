import {Service, Cache as Base} from '@roots/bud-framework'
import crypto from 'crypto'
import {mkdirSync, pathExistsSync, readFileSync} from 'fs-extra'
import {boundMethod as bind} from 'autobind-decorator'
import {sync} from 'globby'
import {isEqual} from 'lodash'

class Cache extends Service implements Base {
  public name = '@roots/bud-cache'

  @bind
  public register(): void {
    this.app.hooks.on('build', config => {
      if (isEqual(config.cache.type, 'memory')) {
        config.cache = {
          type: 'memory',
        }
      }

      return config
    })

    this.app.hooks
      .on('build/cache/name', () => undefined)
      .hooks.on('build/cache/version', () => undefined)
      .hooks.on('build/cache/type', () => 'memory')
      .hooks.on('build/cache/cacheDirectory', () => undefined)
      .hooks.on('build/cache/cacheLocation', () => undefined)
      .hooks.on('build/cache/managedPaths', () => undefined)
  }

  @bind
  public booted(): void {
    this.app.hooks.filter('build/cache/type') == 'filesystem' &&
      !pathExistsSync(this.app.path('storage')) &&
      mkdirSync(this.app.path('storage'))
  }

  @bind
  public version(): string {
    const version = crypto
      .createHash('sha1')
      .update(this.hash())
      .digest('base64')
      .replace(/[^a-z0-9]/gi, '_')
      .toLowerCase()

    return version
  }

  @bind
  public directory(): string {
    return this.app.path('storage', 'cache')
  }

  @bind
  public buildDependencies(): string[] {
    return [
      ...new Set(
        sync([
          this.app.path(
            'project',
            `${this.app.name}.{js,ts,yml,json}`,
          ),
          this.app.path(
            'project',
            `${this.app.name}.config.{js,ts,yml,json}`,
          ),
          this.app.path(
            'project',
            `${this.app.name}.${this.app.mode}.{js,ts.yml,json}`,
          ),
          ...this.app.discovery.resolveFrom.map(
            dep => `${dep}/lib/cjs/index.js`,
          ),
          this.app.path('storage', 'cache/*'),
        ]),
      ),
    ]
  }

  @bind
  public hash(): string {
    return JSON.stringify(
      this.buildDependencies().reduce(
        (all, file) => all.concat(readFileSync(file, 'utf8')),
        process.argv.slice(3).join(''),
      ) ?? '{}',
    )
  }
}

export {Cache}
