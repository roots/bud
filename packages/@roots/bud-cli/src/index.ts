import '@roots/bud'
import '@roots/bud-framework'

import {run} from '@oclif/command'

export {run}
