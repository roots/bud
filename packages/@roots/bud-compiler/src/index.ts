import Compiler from './Compiler'

export {Compiler}
