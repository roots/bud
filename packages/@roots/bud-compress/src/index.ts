import './interface'
import type {Module} from '@roots/bud-framework'

import brotli from './brotli/index'
import gzip from './gzip/index'

const extension: Module = {
  name: '@roots/bud-compress',
  boot: ({extensions}) =>
    [brotli, gzip].forEach(plugin => extensions.add(plugin)),
}

export default extension
export const {name, boot} = extension
