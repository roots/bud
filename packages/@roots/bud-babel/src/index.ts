import './interface'
import type {Framework, Module} from '@roots/bud-framework'
import {Item, Loader} from '@roots/bud-build'
import {Config} from './Config'
import {existsSync} from 'fs-extra'

const babel: Module = {
  name: '@roots/bud-babel',

  api: (app: Framework) => ({
    babel: new Config().init(app),
  }),

  register: ({build, path}) => {
    build.loaders.babel = new Loader(
      require.resolve('babel-loader'),
    )

    build.items.babel = new Item({
      loader: ({build}) => build.loaders.babel,
      options: ({path, babel}) => ({
        cacheDirectory: path('storage', 'cache/babel'),
        root: path('src'),
        presets: Object.values(babel.presets),
        plugins: Object.values(babel.plugins),
      }),
    })

    build.rules.js.setUse(({build}) => [build.items.babel])
  },

  boot: app => {
    !existsSync(app.path('project', 'babel.config.js')) &&
      app.babel.setPresets(['@babel/preset-env']).setPlugins([
        [
          '@babel/plugin-transform-runtime',
          {
            helpers: false,
          },
        ],
        '@babel/plugin-proposal-object-rest-spread',
        '@babel/plugin-syntax-dynamic-import',
        '@babel/plugin-proposal-class-properties',
      ])
  },
}

export default babel
export const {name, api, register, boot} = babel
