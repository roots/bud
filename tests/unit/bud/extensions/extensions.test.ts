import * as extensions from '@roots/bud/src/extensions'

describe('@roots/bud', () => {
  describe('extensions', () => {
    it('is a function`', () => {
      expect(extensions.extensions).toBeInstanceOf(Function)
    })

    it('returns object', () => {
      expect(extensions.extensions()).toMatchSnapshot()
    })
  })
})
