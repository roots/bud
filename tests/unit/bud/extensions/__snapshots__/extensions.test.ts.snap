// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@roots/bud extensions returns object 1`] = `
Object {
  "clean-webpack-plugin": Object {
    "label": "clean-webpack-plugin",
    "make": [Function],
    "options": Object {
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
        "!dll",
      ],
      "cleanStaleWebpackAssets": true,
      "protectWebpackAssets": true,
    },
    "when": [Function],
  },
  "copy-webpack-plugin": Object {
    "label": "copy-webpack-plugin",
    "make": [Function],
    "options": [Function],
    "when": [Function],
  },
  "mini-css-extract-plugin": Object {
    "label": "mini-css-extract-plugin",
    "make": [Function],
    "options": [Function],
    "when": [Function],
  },
  "webpack-define-plugin": Object {
    "label": "webpack-define-plugin",
    "make": [Function],
    "options": [Function],
    "when": [Function],
  },
  "webpack-hot-module-replacement-plugin": Object {
    "label": "webpack-hot-module-replacement-plugin",
    "make": [Function],
    "when": [Function],
  },
  "webpack-manifest-plugin": Object {
    "label": "webpack-manifest-plugin",
    "make": [Function],
    "options": [Function],
    "when": [Function],
  },
  "webpack-provide-plugin": Object {
    "label": "webpack-provide-plugin",
    "make": [Function],
    "when": [Function],
  },
}
`;
